llvm/lib/Target/NVPTX/NVPTXInstrInfo.td

```diff
@@ -3300,6 +3328,13 @@ def : Pat<(bf16 (extractelt (v2bf16 Int32Regs:$src), 1)),
 def : Pat<(v2bf16 (build_vector (bf16 Int16Regs:$a), (bf16 Int16Regs:$b))),
           (V2I16toI32 Int16Regs:$a, Int16Regs:$b)>;
 
+def : Pat<(i16 (extractelt (v2i16 Int32Regs:$src), 0)),
```

Comment:
We're doing exactly the same things for f16/bf16. Perhaps we can just coalesce them all into a loop iterating over the types.

Reply:
you mean just for this pattern or more in general?

Reply:
Specifically for build_vector and extractelt 0/1 for those types. If you see more opportunities, feel free to coalesce them tool. Now that we've switched to i32 registers for all those types, fair number of operations becomes common for those types. I've tried to merge them, but apparently missed some.

Reply:
We still need to specialize based on the scalar type. Is there a way to write a foreach loop that handle this case?

Reply:
At the moment it's a bit cumbersome. You iterate with `foreach foo =  [a,b,c,d] ` over a scalar value. Could be a string with the register type.
Then create an instance of a helper class specialized by it, and which will have fields populated with the values you will need. Somewhat extreme example is WMMA_REGS in llvm/include/llvm/IR/IntrinsicsNVVM.td

We have a pending patch which will add lambdas and functions to tablegen which would make all of this easier, but for now it's awkward. If it turns out to be too much hassle, just place f16/bf16/i16 extractelt/build_vector next to each other, so it's easy to see that they are repeated.

Reply:
ok I coalesced the extract case where the scalar type can be inferred and move the build_vector together.